### PR TITLE
MDS-325 | Added stub v2 endpoints

### PR DESCRIPTION
**Pull request by owen-opencast**

### CODE REVIEW

File: app/uk/gov/hmrc/individualsemploymentsapi/controller/v2/EmploymentsController.scala [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+      val scopes = scopeService.getEndPointScopes("individuals-employments-paye")
+      requiresPrivilegedAuthentication(scopes)
+        .flatMap { authScopes =>
+          throw new Exception("NOT_IMPLEMENTED")
+        }
+        .recover(recovery)
+    }
+  }
+
+  // Home Office and HMCTS want to use the same endpoint,
```
Comment (ajchilts) [inline app/uk/gov/hmrc/individualsemploymentsapi/controller/v2/EmploymentsController.scala:63]: This is a V2 MOJ endpoint. I do not think we should be bringing HO into V2 until we are instructed to do so. Also any such changes should be driven by the handler pattern not on client id. 

HO will remain on V1 for the time being.

File: conf/application.conf [hunk 2]
```diff
@@ -196,3 +196,30 @@ versioning {
   ]
 
 }
+
+api-config {
```
Comment (ajchilts) [inline conf/application.conf:200]: Need to update with the actual config.

File: resources/public/api/conf/P2.0/docs/employments-endpoint-description.md [hunk 3]
```diff
@@ -0,0 +1,3 @@
+<p>This endpoint returns an individual's PAYE employments data for the requested time period.</p>
+<p>The fromDate and toDate are inclusive i.e. the data returned are for the time period that includes the fromDate and the toDate.</p>
+<p>Employment data is only available for tax years commencing 2013-2014.</p>
```
Comment (ajchilts) [inline resources/public/api/conf/P2.0/docs/employments-endpoint-description.md:3]: Is this still true on IF? @JackBrBell - can you please confirm?

File: test/component/uk/gov/hmrc/individualsemploymentsapi/controller/v2/LiveEmploymentsControllerSpec.scala [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+
+    scenario("valid request to the live paye endpoint implementation") {
+      Given("a valid privileged Auth bearer token")
+      AuthStub.willAuthorizePrivilegedAuthToken(authToken, payeEmploymentsScope, retrieveAll = true)
+
+      And("a valid record in the matching API")
+      IndividualsMatchingApiStub.hasMatchingRecord(matchId, nino)
+
+      And("DES will return employments for the NINO")
+      DesStub.searchEmploymentIncomeForPeriodReturns(
```
Comment (ajchilts) [inline test/component/uk/gov/hmrc/individualsemploymentsapi/controller/v2/LiveEmploymentsControllerSpec.scala:180]: This is a V2 endpoint and as such we should not reference DES. Can we please remove? We should not need these DES objects for a not yet implemented endpoint.

File: test/unit/uk/gov/hmrc/individualsemploymentsapi/controller/v2/EmploymentsControllerSpec.scala [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+      new LiveEmploymentsController(
+        mockLiveEmploymentsService,
+        scopeService,
+        mockAuthConnector,
+        hmctsClientId,
+        controllerComponent)
+
+    implicit val hc: HeaderCarrier = HeaderCarrier()
+
+//    when(mockAuthConnector.authorise(any(), eqTo(EmptyRetrieval))(any(), any()))
```
Comment (ajchilts) [inline test/unit/uk/gov/hmrc/individualsemploymentsapi/controller/v2/EmploymentsControllerSpec.scala:75]: Can we remove this?

File: test/unit/uk/gov/hmrc/individualsemploymentsapi/controller/v2/EmploymentsControllerSpec.scala [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+      val result = intercept[Exception] {
+        await(eventualResult)
+      }
+
+      assert(result.getMessage == "NOT_IMPLEMENTED")
+    }
+
+    "fail with status 401 when the bearer token does not have enrolment read:individuals-employments" in new Setup {
+//      when(
+//        mockAuthConnector
```
Comment (ajchilts) [inline test/unit/uk/gov/hmrc/individualsemploymentsapi/controller/v2/EmploymentsControllerSpec.scala:111]: ...and this?

File: test/unit/uk/gov/hmrc/individualsemploymentsapi/controller/v2/EmploymentsControllerSpec.scala [hunk 7]
```diff
@@ -0,0 +1,242 @@
...
+
+      val result = intercept[Exception] {
+        await(res)
+      }
+
+      assert(result.getMessage == "NOT_IMPLEMENTED")
+    }
+
+    "fail with status 401 when the bearer token does not have enrolment read:individuals-employments-paye" in new Setup {
+//      when(
```
Comment (ajchilts) [inline test/unit/uk/gov/hmrc/individualsemploymentsapi/controller/v2/EmploymentsControllerSpec.scala:208]: can we please remove commented out code?

File: test/unit/uk/gov/hmrc/individualsemploymentsapi/controller/v2/EmploymentsControllerSpec.scala [hunk 8]
```diff
@@ -0,0 +1,242 @@
...
+
+      val eventualResult =
+        sandboxEmploymentsController.paye(sandboxMatchId, interval)(FakeRequest())
+
+      val result = intercept[Exception] {
+        await(eventualResult)
+      }
+
+      assert(result.getMessage == "NOT_IMPLEMENTED")
+//      verifyZeroInteractions(mockAuthConnector)
```
Comment (ajchilts) [inline test/unit/uk/gov/hmrc/individualsemploymentsapi/controller/v2/EmploymentsControllerSpec.scala:237]: ... and here?

File: test/component/uk/gov/hmrc/individualsemploymentsapi/VersioningSpec.scala [hunk 9]
```diff
@@ -75,6 +74,19 @@ class VersioningSpec extends BaseSpec {
...
         """)
     }
+
+    scenario("Requests with an accept header version P2.0") {
+
+      When(s"A request to $sandboxMatchEndpointWithSandboxMatchId is made with an accept header for version P2")
+      val response =
+        invokeWithHeaders(sandboxMatchEndpointWithSandboxMatchId, AUTHORIZATION -> authToken, acceptHeaderVP2)
+
+      Then("The response status should be 200")
```
Comment (ajchilts) [inline test/component/uk/gov/hmrc/individualsemploymentsapi/VersioningSpec.scala:84]: Test specifies 200 then we expect INTERNAL_SERVER_ERROR?

File: test/component/uk/gov/hmrc/individualsemploymentsapi/controller/v2/LiveEmploymentsControllerSpec.scala [hunk 10]
```diff
@@ -0,0 +1,221 @@
...
+      Given("a valid privileged Auth bearer token")
+      AuthStub.willAuthorizePrivilegedAuthToken(authToken, employmentsScope, retrieveAll = true)
+
+      And("a valid record in the matching API")
+      IndividualsMatchingApiStub.hasMatchingRecord(matchId, nino)
+
+      When("the root entry point to the API is invoked with a valid match id")
+      val response = invokeEndpoint(s"$serviceUrl/?matchId=$matchId")
+
+      Then("the response status should be 200 (ok)")
```
Comment (ajchilts) [inline test/component/uk/gov/hmrc/individualsemploymentsapi/controller/v2/LiveEmploymentsControllerSpec.scala:103]: Test specifies 200 then we expect INTERNAL_SERVER_ERROR?

File: test/component/uk/gov/hmrc/individualsemploymentsapi/controller/v2/LiveEmploymentsControllerSpec.scala [hunk 11]
```diff
@@ -0,0 +1,221 @@
...
+      DesStub.searchEmploymentIncomeForPeriodReturns(
+        nino,
+        fromDate,
+        toDate,
+        DesEmployments(Seq(DesEmployment(Seq.empty, Some("employer name")))))
+
+      When("the paye endpoint is invoked with a valid match id")
+      val response = invokeEndpoint(s"$serviceUrl/paye?matchId=$matchId&fromDate=$fromDate&toDate=$toDate")
+
+      Then("the response status should be 200 (ok)")
```
Comment (ajchilts) [inline test/component/uk/gov/hmrc/individualsemploymentsapi/controller/v2/LiveEmploymentsControllerSpec.scala:189]: Test specifies 200 then we expect INTERNAL_SERVER_ERROR?

File: test/component/uk/gov/hmrc/individualsemploymentsapi/controller/v2/SandboxEmploymentsControllerSpec.scala [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+        INTERNAL_SERVER_ERROR,
+        "{\"statusCode\":500,\"message\":\"NOT_IMPLEMENTED\"}"
+      )
+    }
+
+    scenario("valid request to the sandbox implementation") {
+      When("I request the root entry point to the API")
+      val response = invokeEndpoint(s"$serviceUrl/sandbox?matchId=57072660-1df9-4aeb-b4ea-cd2d7f96e430")
+
+      Then("The response status should be 200 (ok)")
```
Comment (ajchilts) [inline test/component/uk/gov/hmrc/individualsemploymentsapi/controller/v2/SandboxEmploymentsControllerSpec.scala:76]: Test specifies 200 then we expect INTERNAL_SERVER_ERROR?

Please correct all occurrences.

**Review by ajchilts — CHANGES_REQUESTED** [review]

File: resources/public/api/conf/P2.0/docs/employments-endpoint-description.md (hunk 3 above)
Comment (owen-opencast) [reply resources/public/api/conf/P2.0/docs/employments-endpoint-description.md:3]: I'll remove this for now

**Review by ajchilts — APPROVED** [review]
LGTM 👍

**Comment by mostafasina** [comment]
Regression tests in local dev passed.

**Pull request merged**